lean/bitvec.lean

```diff
@@ -0,0 +1,375 @@
+/-
+Copyright (c) 2015 Joe Hendrix. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joe Hendrix, Sebastian Ullrich, Jason Dagit
+
+Basic operations on bitvectors.
+
+This is a work-in-progress, and contains additions to other theories.
+-/
+import data.vector
+import .nat.lemmas
+
+import tactic.find
+
+import .tactic -- for dec_trivial_tac
+
+-- A `bitvec n` is a subtype of natural numbers such that the value of
+-- the bitvec is strictly less than 2^n.
+@[reducible]
+def bitvec (sz:ℕ) :=
+  { x : ℕ // x < (2 ^ sz) }
+
+namespace bitvec
+open nat
+open vector
+
+section zero
+
+  -- Create a zero bitvector
+  protected
+  def zero (n : ℕ) : bitvec n :=
+    ⟨0, nat.pos_pow_of_pos n dec_trivial⟩
+
+  -- bitvectors have a zero, at every length
+  instance {n:ℕ} : has_zero (bitvec n) := ⟨bitvec.zero n⟩
+
+  @[simp]
+  lemma bitvec_zero_zero (x : bitvec 0) : x.val = 0 :=
+    begin
+      cases x,
+      { simp [nat.pow_zero, lt_one_is_zero] at x_property,
+        simp, assumption
+      }
+    end
+
+end zero
+
+section one
+
+  -- Create a bitvector with the constant one, doing so requires
+  -- the bitvector to have positive length.
+  protected
+  def one_of_pos_len (n: ℕ) {H : n > 0} : bitvec n :=
+    ⟨1, calc
+          1   < 2^1 : by dec_trivial_tac
+          ... ≤ 2^n : by apply (pow_le_pow_of_le_right (zero_lt_succ _) (succ_le_of_lt H))⟩
+
+  -- In pratice, it's more useful to allow 0 length bitvectors to have 1
+  -- as well. This leads to a special case where the 0-length bitvector
+  -- 1 is really just 0. This turns out to simplify things.
+  protected
+  def one (n:ℕ) : bitvec n :=
+    begin
+      cases n,
+      case nat.zero
+      { apply bitvec.zero },
+      case nat.succ
+      { apply bitvec.one_of_pos_len, apply zero_lt_succ }
+    end
```

Comment:
I'm not a fan of using tactics to generate definitions unless they can be avoided.  It make make it hard to read the definition when doing future proofs.  I think this should be defined with a standard match like:

```
protected def one : Π(n:ℕ), bitvec n
| 0 := bitvec.zero 0
| (nat.succ n) := ⟨1, ..⟩
```

In proofs, you usually don't care about the structure of the proof beyond readability, but in definitions, it matters.